A generic dataflow analyzer, with application to liveness analysis

xavierleroy opened the pull request:
[ I'm not sure yet how far to push this approach, ~~so this is marked as "Draft",~~ but high-level reviews and comments are most welcome. ]

There are several dataflow analyses in the ocamlopt compiler back-end.  This PR explores the definition and use of generic dataflow analysis engines to share code between the various analyses and to make these analyses more efficient.

Currently, I have a generic engine for backward dataflow analyses, parameterized by a domain (the type of things inferred by analysis) and a transfer function (the effect of the analysis on individual instructions).  It is used to re-implement the liveness analysis pass.

The generic engine implements a more efficient computation of fixed points arising from loops.  The original liveness analysis takes fixed points starting from bottom, which causes behaviors exponential in the nesting of loops.  The generic engine restarts fixed point iteration for a given loop from the result of the latest iteration of the analysis of the enclosing loop.  This gives the same results if the transfer function is monotonic, but is linear in the nesting of loops.  (See example below.)

The only use of the generic engine so far is liveness analysis.  However, the safe points insertion pass (#10039) may end up using two new analyses of the backward dataflow kind.  These analyses could use the generic engine, possibly with the improvements discussed next.

Possible improvement 1:  use CPS in the engine to run in constant stack space.  Currently we're using stack space roughly proportional to the code size of the function.  (But we're no worse than the current liveness analysis.)

~~Possible improvement 2: make the transfer function lazy in the analysis of the normal and exceptional successors, so that we can stop analyzing at some instructions when there is no need to analyze the successors.  The analyses for safe points insertion would benefit from this optimization.~~  (Done for a special case, should be enough.)

~~Possible improvement 3: find a presentation that is both CPS and lazy, as in 1 + 2.~~

Possible improvement 4: add a generic dataflow engine for forward analyses.

Possible improvement 5: understand how to apply these engines to passes like reloading and spilling, which analyze and transform the code on the fly.

xavierleroy committed:
Add a generic backward dataflow analyzer and use it for liveness analysis

The analyzer is parameterized by an abstract domain and a transfer function.

For recursive handlers, it remembers the latest inferred abstract state
and uses it to start the next fixpoint iteration.  This avoids behaviors
exponential in the nesting of recursive handlers, like we would have
if we started every iteration with bottom.

This exponential behavior was present in the old implementation of liveness
analysis.  It is gone in the new implementation that just calls into
the generic analyzer.

xavierleroy commented:
For the record, here is a function for which the current liveness analysis takes about 1 second on my machine and the new analysis takes negligible time.
```
let iter f =
  for i1 = 0 to 1 do
  for i2 = 0 to 1 do
  for i3 = 0 to 1 do
  for i4 = 0 to 1 do
  for i5 = 0 to 1 do
  for i6 = 0 to 1 do
  for i7 = 0 to 1 do
  for i8 = 0 to 1 do
  for i9 = 0 to 1 do
  for i10 = 0 to 1 do
  for i11 = 0 to 1 do
  for i12 = 0 to 1 do
  for i13 = 0 to 1 do
  for i14 = 0 to 1 do
  for i15 = 0 to 1 do
  for i16 = 0 to 1 do
    ignore (f (i1 + i2 + i3 + i4 + i5 + i6 + i7 + i8 + i9 + i10 + i11))
  done
  done
  done
  done
  done
  done
  done
  done
  done
  done
  done
  done
  done
  done
  done
  done
```
If you add or remove for loops, you'll see the exponential behavior of the current liveness analysis.

gasche commented:
> Possible improvement 2: make the transfer function lazy in the analysis of the normal and exceptional successors, so that we can stop analyzing at some instructions when there is no need to analyze the successors. The analyses for safe points insertion would benefit from this optimization.

Would it not be enough to have the domain values itself be lazy thunks, with `transfer` immediately returning a suspension? Inside the suspension, each transfer case would force `at_next` and `at_raise` as required.

Another naive approach (wondering out loud): in the generic analysis, before matching on `i.desc`, you could call `transfer i D.bot D.bot` and check whether this is `D.top`. In this case there is no need to make any recursive call for this instruction. If I understand correctly, this is less flexible/expressive in general than full laziness, but it suffices if the use of laziness is "cut the computation if you reach Top immediately on some operations".

gasche commented:
> Would it not be enough to have the domain values itself be lazy thunks, with `transfer` immediately returning a suspension? Inside the suspension, each transfer case would force `at_next` and `at_raise` as required.

Actually it would not be enough, as the generic computation itself would not be delayed, so its stack usage remains exactly the same. (The beauty of naive questions for which we figure out the answer immediately *after* asking them.)

xavierleroy commented:
Yes, you're on the right tracks.  But I should have given an example.  For poll point insertion, one of the analyses determines whether all paths cross an `Ialloc` or `Ipoll` instruction.  The abstract domain is `bool` and the transfer function has
```
transfer (Ialloc | Ipoll) at_next at_raise = true
```
So, if we have a path with 10 instructions, then `Ialloc`, then 1000 instructions, maybe we don't need not analyze the 1000 instructions to determine the `at_next` for the `Ialloc`, which the transfer function ignores!

On the other hand, maybe we need to analyze those instructions because they can contain loops which must be analyzed!  So the shortcut may be not applicable here.

xavierleroy commented:
After thinking about it some more, the "shortcut" (skip the analysis of the next instruction if the transfer function is constant for the current instruction) applies to one of the two static analyses involved in poll point insertion, but not in the other.  So it is still worthwhile to support this shortcut.

The latest commit adds a `constant_transfer` optional argument to provide a predicate over operations saying when the analysis of the next instruction can be skipped.

With this, I think the PR is ready for review.  I think we should use this dataflow engine for the new poll point insertion pass, so either this PR is reviewed quickly or I'll have to submit the dataflow engine as part of #10039.

gasche commented:
`constant_transfer` makes a lot of sense. Another option would have been to have `transfer` return a value of type

```ocaml
type transfer =
  | Constant of Dom.t
  | Reader of (end_:Dom.t -> exn:Dom.t -> Dom.t)
```

This would make it easier to interleave the domain computation with the constant-or-not decision, but it also makes the code slightly more complex and the extra expressivity is probably not worth it in most cases.

mshinwell commented:
@xavierleroy I and/or one of my colleagues will review this as soon as possible.

xavierleroy commented:
Thank you @mshinwell !

One thing I should have mentioned to make review easier: this code relies crucially on the uniqueness of catch handler labels within one function.  The previous liveness analysis supported the reuse of labels, following a static scoping discipline.

Octachron commented on asmcomp/dataflow.mli:
```diff
@@ -0,0 +1,98 @@
+     - [i] is a sub-instruction of [instr].
+     - [at_next] is the abstract state "after" the instruction for
+       to normal control flow, falling through the successor(s) of [i].
+     - [at_next] is the abstract state "after" the instruction for
```
at_next ⇒ at_raise

xavierleroy replied:
Well spotted, thanks.

gretay-js commented on asmcomp/dataflow.mli:
```diff
@@ -0,0 +1,99 @@
+     [constant_transfer] returns [true], the [at_next] abstract state
+     is not computed, and the transfer function is called with [D.bot]
+     for the [at_next] argument.
+     A consequence of this optimization is that some reachable parts
+     of the Mach code are not analyzed at all.  For this reason, the
+     second part of the result (the function [int -> D.t] associating
+     an abstract state to each catch handler label) must not be used,
+     as some handlers have not been analyzed.
```
With `constant_transfer` optimization, the abstract value at the entry point (i.e., the first part of the result of [analyze]) is sound, but abstract values associated with any other program points must not be used. This is not only for catch handlers in the second part of the result, but also for any mutable data structure used by the domain.

I think this might be misleading for a generic dataflow analysis and wonder if it makes much difference in practice for poll point insertion.

mshinwell replied:
Having spoken to @gretay-js about her concerns with `constant_transfer` (which I share), I wonder whether we could review and merge the version of this PR without this optimisation first, which should give equal or better performance on liveness analysis compared to the current compiler.  Then we could implement safe points using the new framework and make further decisions (about this optimisation and/or others) based on that experience.  In general, I find it much easier to optimise existing correct code, rather than combining the original implementation with optimisation!

xavierleroy commented:
I hear the concern of @gretay-js and @mshinwell .  On the one hand, this could be addressed by even more detailed documentation comments.  On the other hand, the `constant_transfer` business can be viewed as premature optimization.  

So, I removed the `constant_transfer` optimization (it's archived at https://github.com/xavierleroy/ocaml/tree/dataflowplus) and squashed auxiliary commits.  Let me know if you're happy with the current code.  Meanwhile I'll run a round of CI precheck.

gretay-js reviewed (approved):
Thank you for making this change. I've reviewed the code and it looks good to me.

mshinwell commented:
Indeed, thanks for writing this, which should be a significant improvement.

I have just a couple of stylistic points which I think might improve the interface:
- Add a label for at least one of the arguments of `transfer`.  Maybe we could just use `~exn`, matching the variable name currently in use in the analysis.
- Make the return type into a record, whose labels might be self-descriptive enough that we could shorten the comment.

Also could we indent the code that is under the functor?

xavierleroy committed:
Add labels to the transfer function for dataflow analyses

xavierleroy commented:
>   Add a label for at least one of the arguments of transfer. Maybe we could just use ~exn, matching the variable name currently in use in the analysis.

Right.  There are two arguments of transfer with the same type... so I labeled them both (`~next` and `~exn`)

>   Make the return type into a record, whose labels might be self-descriptive enough that we could shorten the comment.

I see no real need.  The two results have clearly different types, so no confusion is possible.

> Also could we indent the code that is under the functor?

The functor is almost the whole file, so, no, thank you.

xavierleroy commented:
I have other work to submit that builds on top of this PR, so I'm taking the liberty to merge.  Thanks to all for the quick feedback!
